src/backend/distributed/utils/metadata_cache.c

```diff
@@ -298,14 +367,152 @@ LookupDistTableCacheEntry(Oid relationId)
 		cacheEntry->partitionKeyString = partitionKeyString;
 		cacheEntry->partitionMethod = partitionMethod;
 		cacheEntry->shardIntervalArrayLength = shardIntervalArrayLength;
-		cacheEntry->shardIntervalArray = shardIntervalArray;
+		cacheEntry->sortedShardIntervalArray = sortedShardIntervalArray;
+		cacheEntry->shardIntervalCompareFunction = shardIntervalCompareFunction;
+		cacheEntry->hashFunction = hashFunction;
+		cacheEntry->hasUninitializedShardInterval = hasUninitializedShardInterval;
+		cacheEntry->hasUniformHashDistribution = hasUniformHashDistribution;
 	}
 
 	return cacheEntry;
 }
 
 
 /*
+ * ShardIntervalCompareFunction returns the appropriate compare function for the
+ * partition column type. In case of hash-partitioning, it always returns the compare
+ * function for integers. Callers of this function has to ensure that shardIntervalArray
+ * has at least one element.
+ */
+static FmgrInfo *
+ShardIntervalCompareFunction(ShardInterval **shardIntervalArray, char partitionMethod)
```

Comment:
My idea here was to set this function as `static` and allow accesses to `shardIntervalCompareFunction` via the cache. So, I feel like keeping it as it is should be fine.